File: src/indexer.py
from tree_sitter import Language, Parser
import tree_sitter_python
import hashlib
import json
import os

class Indexer:
    """
    Component: The Source of Truth (Indexer)
    Mechanism: AST-Based Semantic Chunking
    """
    def __init__(self, target_schema: str = "python"):
        """
        Tools: ["tree-sitter", "python-bindings"]
        Logic: Initialize Parser with language grammar.
        """
        # Load the Python language grammar
        try:
            self.language = Language(tree_sitter_python.language())
            self.parser = Parser(self.language)
        except Exception as e:
            print(f"Error initializing Tree Sitter: {e}")
            self.parser = None
        
        self.index = {}
        self.current_file = None  # Track current file being parsed

    def parse_file(self, file_path: str) -> dict:
        """
        Task ID: INDEXER_01
        Logic:
          - Parse raw source code into Abstract Syntax Trees (AST).
          - Extract semantic nodes: Function definitions, Class declarations, Structs.
          - Hash each node to create a unique Chunk ID.
          - Map every significant token (function names, variable names) to a Set of Chunk IDs.
        Returns:
          Dictionary of {function_name: source_code_string}
        """
        if not self.parser:
            raise RuntimeError("Parser not initialized.")

        with open(file_path, "rb") as f:
            source_code = f.read()
        
        # Extract module name from file path (for imports)
        self.current_file = os.path.splitext(os.path.basename(file_path))[0]

        tree = self.parser.parse(source_code)
        
        # Helper to traverse and extract functions/classes
        # We want to capture the full source code of the node
        
        extracted_chunks = {}
        
        # Query to find match function/class definitions
        query_scm = """
        (function_definition
          name: (identifier) @name) @function

        (class_definition
          name: (identifier) @name) @class
        """
        
        # Note: tree-sitter query API varies slightly by version, adapting to standard usage
        if not hasattr(self, 'language'):
             self.language = Language(tree_sitter_python.language())
        
        # Use Query constructor as per deprecation warning if possible, but language.query is still working (returning Query)
        # The issue is execution.
        query = self.language.query(query_scm)
        
        from tree_sitter import QueryCursor
        cursor = QueryCursor(query)
        captures = cursor.captures(tree.root_node)

        # captures is a dict {name: [nodes]}
        for capture_name, nodes in captures.items():
            for node in nodes:
                # Logic for processing node
                if capture_name == "name":
                    pass # Handled in parent
                
                if capture_name in ["function", "class"]:
                    # Extract the source code
                    start_byte = node.start_byte
                    end_byte = node.end_byte
                    chunk_source = source_code[start_byte:end_byte].decode("utf8")
                    
                    # Extract name
                    name_node = node.child_by_field_name("name")
                    if name_node:
                        func_name = source_code[name_node.start_byte:name_node.end_byte].decode("utf8")
                        
                        # SPEC COMPLIANCE: Extract function signature
                        signature = {"params": [], "returns": None}
                        
                        # Extract parameters from the 'parameters' field
                        params_node = node.child_by_field_name("parameters")
                        if params_node:
                            # Iterate through children to find identifiers (param names)
                            for child in params_node.children:
                                if child.type == "identifier":
                                    param_name = source_code[child.start_byte:child.end_byte].decode("utf8")
                                    signature["params"].append(param_name)
                                elif child.type == "typed_parameter":
                                    # Handle typed params like `x: int`
                                    name_child = child.child_by_field_name("name") or child.children[0] if child.children else None
                                    if name_child:
                                        param_name = source_code[name_child.start_byte:name_child.end_byte].decode("utf8")
                                        signature["params"].append(param_name)
                                elif child.type == "default_parameter":
                                    # Handle default params like `x=10`
                                    name_child = child.child_by_field_name("name") or child.children[0] if child.children else None
                                    if name_child:
                                        param_name = source_code[name_child.start_byte:name_child.end_byte].decode("utf8")
                                        signature["params"].append(param_name)
                        
                        # Generate ID (hash)
                        chunk_id = hashlib.sha256(chunk_source.encode("utf8")).hexdigest()
                        
                        # Store in index with filename metadata AND signature
                        self.index[func_name] = {
                            "source": chunk_source,
                            "filename": self.current_file,
                            "signature": signature
                        }
                        extracted_chunks[func_name] = chunk_source

        return extracted_chunks

    def export_index(self, output_path: str = "inverted_index.json") -> str:
        """
        Output Artifact: "inverted_index.json"
        """
        with open(output_path, "w") as f:
            json.dump(self.index, f, indent=2)
        return output_path

if __name__ == "__main__":
    # Test stub
    # Create a dummy python file to test
    with open("test_dummy.py", "w") as f:
        f.write("def hello_world():\n    print('Hello')\n\nclass Test:\n    def method(self):\n        pass")
    
    indexer = Indexer()
    extracted = indexer.parse_file("test_dummy.py")
    print("Extracted:", extracted.keys())
    indexer.export_index()
    print("Exported index.")

File: rust_compiler/src/indexer.rs
use anyhow::{Result, Context};
use serde::{Serialize, Deserialize};
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use tree_sitter::{Parser, Query, QueryCursor};

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Chunk {
    pub source: String,
    pub filename: String,
    pub func_name: String,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct Index {
    pub chunks: HashMap<String, Chunk>,
}

pub struct Indexer {
    parser: Parser,
    query: Query,
}

impl Indexer {
    pub fn new() -> Result<Self> {
        let mut parser = Parser::new();
        let language = tree_sitter_python::language();
        parser.set_language(language)
            .map_err(|e| anyhow::anyhow!("Failed to set language: {e}"))?;

        let query_scm = "
        (function_definition
          name: (identifier) @name) @function
        (class_definition
          name: (identifier) @name) @class
        ";
        let query = Query::new(language, query_scm)
            .map_err(|e| anyhow::anyhow!("Failed to compile query: {e}"))?;

        Ok(Self { parser, query })
    }

    pub fn parse_file(&mut self, path: &Path, index: &mut Index) -> Result<()> {
        let source_code = fs::read_to_string(path)
            .with_context(|| format!("Failed to read file: {path:?}"))?;
        
        let filename = path.file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("unknown")
            .to_string();

        let tree = self.parser.parse(&source_code, None)
            .ok_or_else(|| anyhow::anyhow!("Failed to parse code"))?;

        let mut cursor = QueryCursor::new();
        // tree-sitter 0.20 API usage
        let matches = cursor.matches(&self.query, tree.root_node(), source_code.as_bytes());

        for m in matches {
            // Find the @name capture and the @function/@class capture
            let mut func_name = String::new();
            let mut node_byte_range = 0..0;
            
            for capture in m.captures {
                let capture_name = &self.query.capture_names()[capture.index as usize];
                if capture_name == "name" {
                    func_name = capture.node.utf8_text(source_code.as_bytes())?.to_string();
                } else if capture_name == "function" || capture_name == "class" {
                    node_byte_range = capture.node.byte_range();
                }
            }

            if !func_name.is_empty() && node_byte_range.end > node_byte_range.start {
                let chunk_source = &source_code[node_byte_range];
                index.chunks.insert(func_name.clone(), Chunk {
                    source: chunk_source.to_string(),
                    filename: filename.clone(),
                    func_name,
                });
            }
        }
        Ok(())
    }

    pub fn save_index(&self, index: &Index, path: &Path) -> Result<()> {
        let file = fs::File::create(path)?;
        serde_json::to_writer_pretty(file, &index.chunks)?;
        Ok(())
    }
}
